File: src/main/java/com/gitlab/lae/intellij/actions/tree/ui/KeyStrokeLabel.java
package com.gitlab.lae.intellij.actions.tree.ui;

import com.intellij.openapi.keymap.KeymapUtil;

import javax.swing.*;

import static java.awt.event.KeyEvent.*;

final class KeyStrokeLabel {

    static String getKeyText(KeyStroke key) {
        KeyStroke copy = KeyStroke.getKeyStroke(
                VK_A,
                key.getModifiers(),
                key.isOnKeyRelease()
        );
        String prefix = KeymapUtil.getKeystrokeText(copy);
        prefix = prefix.substring(0, prefix.length() - 1);
        String suffix = key.getKeyChar() != CHAR_UNDEFINED
                ? String.valueOf(key.getKeyChar())
                : getKeyText(key.getKeyCode());
        return prefix + suffix;
    }

    private static String getKeyText(int keyCode) {
        switch (keyCode) {
            case VK_MINUS: return "-";
            case VK_COMMA: return ",";
            case VK_QUOTE: return "'";
            default: return KeymapUtil.getKeyText(keyCode);
        }
    }
}

File: src/main/java/com/gitlab/lae/intellij/actions/tree/ui/KeyStrokeLabel.kt
package com.gitlab.lae.intellij.actions.tree.ui

import com.intellij.openapi.keymap.KeymapUtil
import java.awt.event.KeyEvent
import java.awt.event.KeyEvent.CHAR_UNDEFINED
import javax.swing.KeyStroke

fun getKeyText(key: KeyStroke): String {
  val copy = KeyStroke.getKeyStroke(
    KeyEvent.VK_A,
    key.modifiers,
    key.isOnKeyRelease
  )

  var prefix = KeymapUtil.getKeystrokeText(copy)
  prefix = prefix.substring(0, prefix.length - 1)

  val suffix =
    if (key.keyChar != CHAR_UNDEFINED) key.keyChar.toString()
    else getKeyText(key.keyCode)

  return prefix + suffix
}

private fun getKeyText(keyCode: Int): String = when (keyCode) {
  KeyEvent.VK_MINUS -> "-"
  KeyEvent.VK_COMMA -> ","
  KeyEvent.VK_QUOTE -> "'"
  else -> KeymapUtil.getKeyText(keyCode)
}
